common/diagnostics/Metrics.java

```diff
@@ -6,203 +6,761 @@
 
 package com.vaticle.typedb.core.common.diagnostics;
 
+import com.eclipsesource.json.JsonArray;
 import com.eclipsesource.json.JsonObject;
-
-import java.util.TimeZone;
+import com.vaticle.typedb.common.collection.ConcurrentSet;
+import com.vaticle.typedb.core.common.parameters.Arguments;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.lang.management.ManagementFactory;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicLong;
 
+import static com.vaticle.typedb.core.common.diagnostics.StatisticReporter.REPORT_INTERVAL_MINUTES;
+import static com.vaticle.typedb.core.common.diagnostics.Version.JSON_API_VERSION;
+
 public class Metrics {
-    private final SystemProperties system;
-    private final NetworkRequests requests;
-    private final CurrentCounts usage;
-    private final UserErrorStatistics userErrors;
+    private final ConcurrentSet<String> primaryDatabaseHashes = new ConcurrentSet<>();
+    private final BaseProperties base;
+    private final ServerProperties serverProperties;
+    private final ConcurrentMap<String, DatabaseLoadDiagnostics> databaseLoad = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, NetworkRequests> requests = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, UserErrorStatistics> userErrors = new ConcurrentHashMap<>();
+
+    public Metrics(String deploymentID, String serverID, String name, String version, boolean reportingEnabled, Path dataDirectory) {
+        this.base = new BaseProperties(JSON_API_VERSION, deploymentID, serverID, name, reportingEnabled);
+        this.serverProperties = new ServerProperties(version, dataDirectory);
+    }
 
-    public Metrics(String serverID, String name, String version) {
-        this.system = new SystemProperties(serverID, name, version);
-        this.requests = new NetworkRequests();
-        this.usage = new CurrentCounts();
-        this.userErrors = new UserErrorStatistics();
+    public void takeSnapshot() {
+        this.base.updateSinceTimestamp();
+        this.databaseLoad.values().forEach(DatabaseLoadDiagnostics::takeSnapshot);
+        this.requests.values().forEach(NetworkRequests::takeSnapshot);
+        this.userErrors.values().forEach(UserErrorStatistics::takeSnapshot);
     }
 
-    public void requestAttempt(Metrics.NetworkRequests.Kind kind) {
-        requests.attempt(kind);
+    private String hashAndAddDatabaseIfAbsent(@Nullable String databaseName) {
+        String databaseHash = databaseName != null ? String.valueOf(databaseName.hashCode()) : "";
+
+        if (!databaseHash.isEmpty()) {
+            this.databaseLoad.computeIfAbsent(databaseHash, val -> new DatabaseLoadDiagnostics());
+        }
```

Comment:
These should be combined into one `if (databaseName != null)` statement for readability.

Reply:
What do I do with `databaseHash` declaration? Will it be more readable than this version?

Reply:
You can have it uninitialized outside the if-statement, and initialize it in each branch. Java will complain if it's not guaranteed to be initialized at the end of it.

Reply:
Well, if you prefer `if-else` over this, will do.

Reply:
This is a little more dangerous
```java
String databaseHash = "";

if (databaseName != null) {
    // oops
    var databaseHash = String.valueOf(databaseName.hashCode());
    this.databaseLoad.computeIfAbsent(databaseHash, val -> new DatabaseLoadDiagnostics());
}
```
`else databaseHash = "";` will allow the Java compiler to catch that edit.

Reply:
I'm not sure why would anyone want to do it. I'm not strongly against it, but this example could be applied anywhere else. Do we really prefer to have "else" instead of a `default initialization` + one override inside `if`?

Just to understand our basic approach.

Reply:
The thinking is so that it's impossible to forget to initialize the variable to a sensible value rather than always risking leaving the fallback unchanged.

It also reduces complexity. If `databaseHash` is uninitialized before the conditional, and used after, that means each branch of the conditional must have made a decision what to initialize it to. If the conditional is non-exhaustive and the variable has a default value, the reader must inspect each branch to know if and when the default is overridden. (And also be cognizant of shadowing as in the "oops" example).

It's perhaps less relevant here because the snippet is so small, but a nice pattern to follow nonetheless.

Reply:
Okay, will do.